src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -0,0 +1,465 @@
+import com.gs.collections.api.list.ImmutableList;
+import com.gs.collections.api.list.MutableList;
+import com.gs.collections.api.multimap.MutableMultimap;
+import com.gs.collections.api.partition.PartitionIterable;
+import com.gs.collections.impl.block.factory.Functions;
+import com.gs.collections.impl.block.function.checked.CheckedFunction;
+import com.gs.collections.impl.factory.Lists;
+import com.gs.collections.impl.factory.Multimaps;
+import com.gs.collections.impl.list.mutable.FastList;
+import kafka.common.ErrorMapping;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.integration.kafka.core.BrokerAddress;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.ConsumerException;
+import org.springframework.integration.kafka.core.FetchRequest;
+import org.springframework.integration.kafka.core.KafkaConsumerDefaults;
+import org.springframework.integration.kafka.core.KafkaMessage;
+import org.springframework.integration.kafka.core.KafkaMessageBatch;
+import org.springframework.integration.kafka.core.KafkaTemplate;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.Result;
+import org.springframework.util.Assert;
+
+/**
+ * @author Marius Bogoevici
+ */
+public class KafkaMessageListenerContainer implements SmartLifecycle {
+
+	private static final Log log = LogFactory.getLog(KafkaMessageListenerContainer.class);
+
+	public static final Function<Map.Entry<Partition, ?>, Partition> keyFunction = Functions.getKeyFunction();
+
+	private final GetOffsetForPartitionFunction getOffset = new GetOffsetForPartitionFunction();
+
+	private final PartitionToLeaderFunction getLeader = new PartitionToLeaderFunction();
+
+	private final Function<Partition, Partition> passThru = Functions.getPassThru();
+
+	private final LaunchFetchTaskProcedure launchFetchTask = new LaunchFetchTaskProcedure();
+
+	private final Object lifecycleMonitor = new Object();
+
+	private final KafkaTemplate kafkaTemplate;
+
+	private final Partition[] partitions;
+
+	public boolean autoStartup = true;
+
+	private Executor fetchTaskExecutor;
+
+	private Executor adminTaskExecutor = Executors.newSingleThreadExecutor();
+
+	private int concurrency = 1;
+
+	private volatile boolean running = false;
+
+	private int maxFetch = KafkaConsumerDefaults.FETCH_SIZE_INT;
+
+	private MessageListener messageListener;
+
+	private volatile OffsetManager offsetManager;
+
+	private ConcurrentMap<Partition, Long> fetchOffsets;
+
+	private ConcurrentMessageListenerDispatcher messageDispatcher;
+
+	private final MutableMultimap<BrokerAddress, Partition> partitionsByBrokerMap = Multimaps.mutable.set.with();
+
+	public KafkaMessageListenerContainer(ConnectionFactory connectionFactory, Partition[] partitions) {
```

Comment:
Can't `partitions` be a `varargs array` argument? As well as for the second ctor.
To simplify usage from Java